falcon/cors.py

```diff
@@ -0,0 +1,426 @@
+        # this response
+        if self.origins_vary:
+            self._set_vary_origin(resp)
+
+        origin = req.get_header('origin')
+        # 6.1.1
+        # 6.2.1
+        if not origin:
+            return
+
+        # 6.1.2
+        # 6.1.3 (Access-Control-Allow-Origin)
+        # 6.2.2
+        # 6.2.7 (Access-Control-Allow-Origin)
+        if not self._process_origin(req, resp, origin):
+            return
+
+        # Basic or actual request
+        if req.method != 'OPTIONS':
+            # 6.1.3 (Access-Control-Allow-Credentials)
+            self._process_credentials(req, resp, origin)
+
+            # 6.1.4
+            self._process_expose_headers(req, resp)
+        # Preflight request
+        else:
+            request_method = req.get_header('access-control-request-method')
+            # 6.2.3
+            if not request_method:
+                return
+
+            # 6.2.4
+            requested_header_list = self._get_requested_headers(req)
+
+            # 6.2.5
+            # 6.2.9
+            if not self._process_methods(req, resp, resource):
+                return
+
+            # 6.2.6
+            # 6.2.10
+            if not self._process_allow_headers(req, resp, requested_header_list):
+                return
+
+            # 6.2.7 (Access-Control-Allow-Credentials)
+            self._process_credentials(req, resp, origin)
+
+            # 6.2.8
+            self._process_max_age(req, resp)
+
+    def _process_origin(self, req, resp, origin):
+        """Inspects the request and adds the Access-Control-Allow-Origin
+        header if the requested origin is allowed.
+
+        Returns:
+            ``True`` if the header was added and the requested origin
+            is allowed, ``False`` if the origin is not allowed and the
+            header has not been added.
+        """
+        if self._cors_config['allow_all_origins']:
+            if self.supports_credentials:
+                self._set_allow_origin(resp, origin)
+            else:
+                self._set_allow_origin(resp, '*')
+            return True
+
+        if origin in self._cors_config['allow_origins_list']:
+            self._set_allow_origin(resp, origin)
+            return True
+
+        regex = self._cors_config['allow_origins_regex']
+        if regex is not None:
+            if regex.match(origin):
+                self._set_allow_origin(resp, origin)
+                return True
+
+        return False
+
+    def _process_allow_headers(self, req, resp, requested_headers):
+        """Adds the Access-Control-Allow-Headers header to the response,
+        using the cors settings to determine which headers are allowed.
+
+        Returns:
+            True if all the headers the client requested are allowed.
+            False if some or none of the headers the client requested are allowed.
+        """
+        if not requested_headers:
+            return True
+        elif self._cors_config['allow_all_headers']:
```

Comment:
since both if conditions have a return statement, i have been suggested that elif should be replace by an if.